report_templates/entitlements.erb

```diff
@@ -2,17 +2,25 @@
 name: Subscription - Entitlement Report
 snippet: false
 model: ReportTemplate
+template_inputs:
+- name: days_from_now
+  required: false
+  input_type: user
+  description: "Show subscriptions expiring in <= this many days.  Leave blank to show all entitlements."
+  advanced: false
+  value_type: plain
+  options: "7\r\n30\r\n60\r\n90\r\n120\r\n180"
 require:
 - plugin: katello
   version: 3.14.0
 -%>
-<%- report_headers 'Name', 'Organization', 'Lifecycle Environment', 'Content View', 'Host Collections', 'Virtual', 'Guest of Host', 'OS', 
-'Arch', 'Sockets', 'RAM', 'Cores', 'SLA', 'Products', 'Subscription Name', 'Subscription Type', 'Subscription Quantity', 
-'Subscription SKU', 'Subscription Contract', 'Subscription Account', 'Subscription Start', 'Subscription End', 'Subscription Guest' -%>
 <%- load_hosts(includes: [:lifecycle_environment, :operatingsystem, :architecture, :content_view, :organization, :reported_data, :subscription_facet, :pools => [:subscription]]).each_record do |host| -%>
+<%-   days_from_now = input('days_from_now').to_i -%>
+<%-   should_filter = !input('days_from_now').empty? -%>
 <%-   host.pools.each do |pool| -%>
+<%-     next if should_filter && pool.expiring_in_days > days_from_now -%>
```

Comment:
This is filtering during the filter runtime, is there a way to convert this to SQL through a search condition passed to `load_hosts` macro? Notw that it accepts search syntax. This would have big performance benefit especially for people having many hosts.

Reply:
@ares Thanks, updated to use `search` instead of filtering in the erb.
By using scoped_search and a custom external method (see the Katello PR) I am able to get scoped_search to return only hosts _with expiring pools_.  However, when iterating through each host, `host.pools` would still return _all_ of that host's pools, not just the expiring ones.  So, I also added an instance method to `Host`, `pools_expiring_in_days`, so that none of the filtering has to be done inside the report template.